lib/handler/cache-handler.js

```diff
@@ -0,0 +1,348 @@
+'use strict'
+
+const util = require('../core/util')
+const DecoratorHandler = require('../handler/decorator-handler')
+const { parseCacheControlHeader, parseVaryHeader, UNSAFE_METHODS } = require('../util/cache')
+
+/**
+ * Writes a response to a CacheStore and then passes it on to the next handler
+ */
+class CacheHandler extends DecoratorHandler {
+  /**
+   * @type {import('../../types/cache-interceptor.d.ts').default.CacheOptions} | null
+   */
+  #opts = null
+  /**
+   * @type {import('../../types/dispatcher.d.ts').default.RequestOptions | null}
+   */
+  #req = null
+  /**
+   * @type {import('../../types/dispatcher.d.ts').default.DispatchHandlers | null}
+   */
+  #handler = null
+  /**
+   * @type {import('../../types/cache-interceptor.d.ts').default.CacheStoreWriteable | undefined}
+   */
+  #writeStream
+
+  /**
+   * @param {import('../../types/cache-interceptor.d.ts').default.CacheOptions} opts
+   * @param {import('../../types/dispatcher.d.ts').default.RequestOptions} req
+   * @param {import('../../types/dispatcher.d.ts').default.DispatchHandlers} handler
+   */
+  constructor (opts, req, handler) {
+    super(handler)
+
+    this.#opts = opts
+    this.#req = req
+    this.#handler = handler
+  }
+
+  /**
+   * @see {DispatchHandlers.onHeaders}
+   *
+   * @param {number} statusCode
+   * @param {Buffer[]} rawHeaders
+   * @param {() => void} resume
+   * @param {string} statusMessage
+   * @param {Record<string, string | string[]> | undefined} headers
+   * @returns {boolean}
+   */
+  onHeaders (
+    statusCode,
+    rawHeaders,
+    resume,
+    statusMessage,
+    headers = util.parseHeaders(rawHeaders)
+  ) {
+    const downstreamOnHeaders = () => this.#handler.onHeaders(
+      statusCode,
+      rawHeaders,
+      resume,
+      statusMessage,
+      headers
+    )
+
+    if (
+      this.#req.method in UNSAFE_METHODS &&
+      statusCode >= 200 &&
+      statusCode <= 399
+    ) {
+      // https://www.rfc-editor.org/rfc/rfc9111.html#name-invalidating-stored-respons
+      const result = this.#opts.store.deleteByOrigin(this.#req.origin)
+      if (
+        result &&
+        typeof result.catch === 'function' &&
+        typeof this.#handler.onError === 'function'
+      ) {
+        result.catch(err => {
+          process.emitWarning(`deleteByOrigin failed: ${err}`)
```

Comment:
why are we emitting a process warning here?

Reply:
Somewhat similar idea to https://github.com/nodejs/undici/pull/3562#discussion_r1784212558, an error here shouldn't cause the request to fail and this gives some visibility to when these happen

Reply:
The global warning event is not really how you give this visibility. I would rather call a user-provided function when these happens, call a method on the store, or emit an event on the store object.